### PR TITLE
ROS: Don't skip blocks passed end

### PR DESCRIPTION
**Pull request by benaadams**
Seen in https://github.com/dotnet/corefx/pull/28095#discussion_r175247253

/cc @AlexRadch @pakrym @davidfowl @ahsonkhan

### CODE REVIEW

File: src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs [hunk 1]
```diff
@@ -181,6 +181,37 @@ public void SeekEmptySkipDoesNotCrossPastEnd()
 
             Assert.Equal(0, c1.GetInteger());
             Assert.Equal(bufferSegment2, c1.GetObject());
+
+            // Go out of bounds for segment
+            Assert.Throws<ArgumentOutOfRangeException>(() => c1 = buffer.GetPosition(150, buffer.Start));
+            Assert.Throws<ArgumentOutOfRangeException>(() => c1 = buffer.GetPosition(250, buffer.Start));
```
Comment (benaadams) [inline src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs]: These asserts currently work; just adding for completeness

File: src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs [hunk 2]
```diff
@@ -181,6 +181,37 @@ public void SeekEmptySkipDoesNotCrossPastEnd()
...
+            Assert.Equal(bufferSegment2, c1.GetObject());
+
+            c1 = buffer.GetPosition(100, buffer.Start);
+
+            Assert.Equal(0, c1.GetInteger());
+            Assert.Equal(bufferSegment2, c1.GetObject());
+
+            // Go out of bounds for segment
+            Assert.Throws<ArgumentOutOfRangeException>(() => c1 = buffer.GetPosition(150, buffer.Start));
+            Assert.Throws<ArgumentOutOfRangeException>(() => c1 = buffer.GetPosition(250, buffer.Start));
```
Comment (benaadams) [inline src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs]: These asserts are where it currently breaks as it skips the blocks without checking if its already hit the end

**Comment by davidfowl** [comment]
@Petermarcu @karelz @joshfree We want this in preview2, do we need to change the branch?

**Comment by joshfree** [comment]
> @Petermarcu @karelz @joshfree We want this in preview2, do we need to change the branch?

Getting this into master is still fine for preview2 at this point, I believe.  There should be another mop up integration to the release branch after the bug bounce.

@leecow it would be great to publish the schedule soon 😄

File: src/System.Memory/src/System/Buffers/ReadOnlySequence_helpers.cs [hunk 3]
```diff
@@ -275,7 +275,7 @@ private static SequencePosition SeekMultiSegment(ReadOnlySequenceSegment<T> curr
 
                     // Move to next
                     count -= memoryLength;
-                    currentSegment = currentSegment.Next;
+                    currentSegment = isCurrentAtEnd ? null : currentSegment.Next;
```
Comment (pakrym) [inline src/System.Memory/src/System/Buffers/ReadOnlySequence_helpers.cs]: This is not the bug

**Comment by pakrym** [comment]
https://github.com/pakrym/corefx/commit/9dc3fbda05940222ac8f82ab12ca1e3258c863e5#diff-189d999cfee30003f64a57258ee0451e

**Comment by pakrym** [comment]
The bug we are seeing is because of https://github.com/dotnet/corefx/blob/master/src/System.Memory/src/System/Buffers/ReadOnlySequence_helpers.cs#L155 this check that looks at `next.Memory.Length` which could change from under readonly sequence returning different results for enumeration.

**Comment by leecow** [comment]
@joshfree - it's coming any minute now :-P

**Comment by benaadams** [comment]
Alpine and OSX are
```
TypeError: cannot concatenate 'str' and 'type' objects
```
Which is a buildtools issue?

**Comment by ahsonkhan** [comment]
> Which is a buildtools issue?

Yes, @Anipik fixed the issue in https://github.com/dotnet/buildtools/pull/1969.

**Comment by benaadams** [comment]
This PR is be superseded by https://github.com/dotnet/corefx/pull/28294; is that correct @ahsonkhan ?

**Pull request closed without merging**